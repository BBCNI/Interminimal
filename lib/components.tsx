import React, {
  ComponentType,
  createContext,
  createElement,
  ReactNode,
  Children,
  useContext,
  Fragment,
  cloneElement,
  isValidElement,
  forwardRef,
  ReactElement,
  Ref
} from "react";

import {
  AsType,
  TProps,
  TDictionaryRoot,
  TextPropType,
  TranslateLocalProps,
  TranslateProps
} from "./types";

import { parseTemplate } from "./template";
import { LangContext } from "./context";
import { TString } from "./string";

const TContext = createContext(new LangContext());

/**
 * Hook that gets the currently active translation context. Here's an example
 * of a component that wraps the `Intl.DateTimeFormat` API.
 *
 * ```typescript
 * const TDateFormat: ComponentType<{ date: Date }> = ({ date }) => {
 *   // Get the context
 *   const ctx = useTranslation();
 *   // Use context's languages stack to find a format for our locale
 *   const dtf = new Intl.DateTimeFormat(ctx.languages);
 *   // Find out which language was matched...
 *   const { locale } = dtf.resolvedOptions();
 *   const ts = TString.literal(dtf.format(date), locale);
 *   return <T text={ts} />;
 * };
 * ```
 *
 * @returns the active translation context
 * @category Hooks
 */
export const useTranslation = (): LangContext => useContext(TContext);

/**
 * Wrap components in a nested [[`LangContext`]]. Used to override settings in
 * the context. For example we can add an additional dictionary.
 *
 * ```typescript
 * const Miscount = ({ children }: { children: ReactNode }) => {
 *   // pretend one is three
 *   const dict = { $$dict: { one: { en: "three" } } };
 *   return <TranslateLocal dictionary={dict}>{children}</TranslateLocal>;
 * };
 * ```
 * @category Components
 */
export const TranslateLocal: ComponentType<TranslateLocalProps> = ({
  children,
  ...props
}): ReactElement => {
  const ctx = useTranslation().derive(props);
  return <TContext.Provider value={ctx}>{children}</TContext.Provider>;
};

/**
 * Wrap components in a nested [[`LangContext`]] that establishes a new
 * language stack. By default any children will be wrapped in a `div` with
 * a `lang=` property that indicates the language of the wrapped content.
 *
 * Within this context any content which can't be translated into the requested
 * languages will have it's own `lang=` property to reflect the fact that it
 * is in a different language than expected.
 *
 * ```typescript
 * // Renders as <div lang="cy">....</div>
 * const Welsh: ComponentType<{ children: ReactNode }> = ({ children }) => (
 *   <Translate lang="cy">{children}</Translate>
 * );
 *
 * // Renders as <section lang="cy">....</section>
 * const WelshSection: ComponentType<{ children: ReactNode }> = ({ children }) => (
 *   <Translate as="section" lang="cy">
 *     {children}
 *   </Translate>
 * );
 * ```
 *
 * Unlike [[`TranslateLocal`]] `Translate` always wraps the translated
 * content in an element with a `lang=` property.
 *
 * @category Components
 */
export const Translate: ComponentType<TranslateProps> = ({
  children,
  as = "div",
  ...props
}): ReactElement => {
  const ctx = useTranslation().derive(props);
  return (
    <TText as={as} lang={ctx.language}>
      <TContext.Provider value={ctx}>{children}</TContext.Provider>
    </TText>
  );
};

interface AsProps {
  as: AsType;
  children?: ReactNode;
  [key: string]: any;
}

const As: ComponentType<AsProps> = forwardRef<ReactElement, AsProps>(
  ({ as, children, ...props }, ref) =>
    createElement(as, { ref, ...props }, children)
);

As.displayName = "As";

interface TTextProps {
  children: ReactNode;
  lang: string;
  as: AsType;
  [key: string]: any;
}

const TText: ComponentType<TTextProps> = forwardRef<ReactElement, TTextProps>(
  ({ children, lang, as, ...props }, ref): ReactElement => {
    const ctx = useTranslation();

    if (lang !== ctx.ambience)
      return (
        <TranslateLocal ambient={lang}>
          <As as={as} ref={ref} {...props} lang={lang}>
            {children}
          </As>
        </TranslateLocal>
      );
    return (
      <As as={as} ref={ref} {...props}>
        {children}
      </As>
    );
  }
);

TText.displayName = "TText";

interface TFormatProps {
  format: string;
  lang: string;
  children: ReactNode;
  ref?: Ref<ReactElement>;
}

const TFormat: ComponentType<TFormatProps> = forwardRef<
  ReactElement,
  TFormatProps
>(({ format, lang, children }, ref): ReactElement => {
  const clone = (elt: ReactNode, props?: any): ReactNode => {
    if (isValidElement(elt)) return cloneElement(elt, props);
    if (process.env.NODE_ENV !== "production")
      throw new Error(`Can't add props to a non-element`);
  };

  const parts = parseTemplate(format);

  // Bail out quickly in the simple case
  if (parts.length === 1 && typeof parts[0] === "string")
    return <Fragment>{parts[0]}</Fragment>;

  // Make children into a regular array of nodes
  const params = Children.toArray(children);

  if (process.env.NODE_ENV !== "production")
    if (ref && params.length !== 1)
      // Passing a ref is a special case which only allows
      // a single child
      throw new Error(`Can only forward refs to single children`);

  // Set of available indexes
  const avail = new Set(params.map((_x: any, i: number) => i + 1));

  const dict: TDictionaryRoot = { $$dict: {} };

  // Output nodes
  const out = parts.map(part => {
    if (typeof part === "string") return part;
    const { index, name, text } = part;

    if (name && text)
      dict.$$dict[name] = TString.literal(text, lang).dictionary;

    if (index < 1 || index > params.length)
      throw new Error(
        `Arg out of range %${index} (1..${params.length} are valid)`
      );

    if (!avail.has(index)) throw new Error(`Already using arg %${index}`);

    // Mark it used
    avail.delete(index);

    // If we're passing a ref clone it in. Only do this to the first
    // parameter. This check pretty redundant - there's a check above
    // that enforces singularity in the ref passed case.
    if (ref && index === 1) return clone(params[index - 1], { ref });
    return params[index - 1];
  });

  if (process.env.NODE_ENV !== "production")
    if (avail.size) throw new Error(`Unused args: ${avail}`);

  if (Object.keys(dict.$$dict).length)
    return <TranslateLocal dictionary={dict}>{out}</TranslateLocal>;

  return <Fragment>{out}</Fragment>;
});

TFormat.displayName = "TFormat";

const noRef = (ref: Ref<ReactElement>) => {
  if (ref) throw new Error(`Can't pass ref`);
};

function resolveTranslationProps(
  ctx: LangContext,
  tag?: string,
  text?: TextPropType
): TString {
  const r = () => {
    if (process.env.NODE_ENV !== "production")
      if (tag && text) throw new Error(`Got both tag and text`);
    if (text) return ctx.resolve(text);
    if (tag) return ctx.resolve([tag]);
    // istanbul ignore next - can't happen
    throw new Error(`No text or tag`);
  };
  return r().toLang(ctx.languages);
}

/**
 * A wrapper for content that should be translated. It attempts to translate
 * the content you give it according to the active [[`LangContext`]]. It can
 * translate content looked up in the translation dictionary and fat strings
 * (or [[`TString`]]s).
 *
 * It can optionally perform template substitution on the translated text,
 * allowing child components to render portions of the translated text
 * with arbitrary wrappers.
 *
 * By default translated text is wrapped in a `span`. Render a different
 * element using the `as` property.
 *
 * If the wrapped content can't be translated into the context's preferred
 * language it will have a `lang=` property specifying its actual language.
 *
 * The simplest usage is to render translatable content without template
 * substition:
 *
 * ```typescript
 * // Render multilingual content
 * const hi = { en: "Hello", de: "Hallo", fr: "Bonjour" };
 * return <T content={hi} />;
 * // fr: <span>Bonjour</span>
 * ```
 *
 * Template substitution allows you to build whole component trees from a
 * translated string:
 *
 * ```typescript
 * const info = {
 *   en: "Here's a %1[useful link] and here's some %2[italic text]",
 *   fr: "Voici %2[du texte en italique] et un %1[lien utile]",
 *   de: "Hier ist ein %1[n√ºtzlicher Link] und hier ein %2[kursiver Text]"
 * };
 * return (
 *   <T as="div" text={info}>
 *     <T as="a" tag="%1" href="/" />
 *     <T as="i" tag="%2" />
 *   </T>
 * );
 * // fr:
 * //   <div>
 * //     Voici <i>du texte en italique</i> et un <a href="/">lien utile</a>
 * //   </div>
 * ```
 *
 * You can also look up and translate dictionary tags:
 *
 * ```typescript
 * // Same as the previous example if `info` is in the dictionary
 * return (
 *   <T as="div" tag="info">
 *     <T as="a" tag="%1" href="/" />
 *     <T as="i" tag="%2" />
 *   </T>
 * );
 * ```
 *
 * @category Components
 */
export const T: ComponentType<TProps> = forwardRef<ReactElement, TProps>(
  (
    { children, tag, text, content, count, as = "span", ...props },
    ref
  ): ReactElement => {
    const ctx = useTranslation();

    if (content) {
      if (process.env.NODE_ENV !== "production") {
        if (tag || text)
          throw new Error(`Please don't mix content with tag or text`);
        noRef(ref);
      }

      const ts = ctx.translate(content);
      return (
        <TText
          as={as}
          lang={ts.language}
          {...ctx.resolveMagicProps(props, ts.language)}
        >
          {ts.toString(count)}
        </TText>
      );
    }

    if (tag || text) {
      const ts = resolveTranslationProps(ctx, tag, text);

      return (
        <TText
          as={as}
          lang={ts.language}
          {...ctx.resolveMagicProps(props, ts.language)}
        >
          <TFormat ref={ref} lang={ts.language} format={ctx.render(ts, count)}>
            {children}
          </TFormat>
        </TText>
      );
    }

    if (process.env.NODE_ENV !== "production") noRef(ref);

    return (
      <TText as={as} lang={ctx.defaultLang} {...ctx.resolveMagicProps(props)}>
        {children}
      </TText>
    );
  }
);

T.displayName = "T";

const boundMap = new Map();

/**
 * Create a new component that behaves like `<T>` but with a different default
 * `as` element.
 *
 * ```typescript
 * const Toption = tBind("option");
 * // later
 * return <Toption value="1" tag="one" />
 * ```
 *
 * It's also possible to wrap React components.
 *
 * ```typescript
 * const TImage = tBind(Image as FunctionComponent);
 * ```
 *
 * The need for the cast is ugly. Not sure how to fix that. PRs welcome...
 *
 * The generated components are cached - so whenever you call `tBind("p")` you
 * will get the same component.
 *
 * @category Utilities
 */
export const tBind = (as: AsType): ComponentType<TProps> => {
  const bind = (as: AsType): ComponentType<TProps> => {
    const bound: ComponentType<TProps> = forwardRef(
      ({ children, ...props }, ref) => (
        <T as={as} ref={ref} {...props}>
          {children}
        </T>
      )
    );
    const asName = typeof as === "string" ? as : as.displayName;
    if (asName) {
      bound.displayName = `T${asName}`;
      Object.defineProperty(bound, "name", { value: bound.displayName });
    }
    return bound;
  };

  let bound = boundMap.get(as);
  if (!bound) boundMap.set(as, (bound = bind(as)));
  return bound;
};

/**
 * Make multiple bound versions of `<T>` at once.
 *
 * ```typescript
 * const [Tli, Tdiv, Th2, Tp] = tBindMulti(["li", "div", "h2", "p"]);
 * ```
 *
 * @category Utilities
 */
export const tBindMulti = (as: AsType[]) => as.map(tBind);
