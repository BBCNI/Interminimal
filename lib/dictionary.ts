import { TFatString, TString } from "./string";
import { NextCache } from "./nextCache";

type TDictionaryType = {
  [key: string]: TFatString | TDictionaryRoot;
} & { $$dict?: never };

/**
 * @category Dictionary
 */
export type TDictionaryMeta = {
  [key: string]: string;
};

/**
 * The root of a translation dictionary
 *
 * @category Dictionary
 */
export type TDictionaryRoot = {
  /** the translations */
  $$dict: TDictionaryType;
  /** optional metadata - ignored by _Interminimal_ */
  $$meta?: TDictionaryMeta;
};

/**
 * @category Dictionary
 */
export const checkDictionary = (dictionary: TDictionaryRoot): void => {
  if (!("$$dict" in dictionary))
    throw new Error(`Invalid dictionary (missing $$dict key)`);

  Object.values(dictionary.$$dict).map(ts =>
    "$$dict" in ts ? checkDictionary(ts as TDictionaryRoot) : TString.cast(ts)
  );
};

const canMerge = (obj: any): boolean =>
  obj && typeof obj === "object" && !Array.isArray(obj);

// Frozen dictionary merge
const mergeObj = (a: any, b: any): any => {
  if (canMerge(a) && canMerge(b))
    return Object.fromEntries(
      [...new Set([...Object.keys(a), ...Object.keys(b)])].map(key => {
        if (key in b) {
          if (key in a) return [key, merge(a[key], b[key])];
          return [key, b[key]];
        }
        return [key, a[key]];
      })
    );
  return b;
};

const merge = (a: any, b: any) => Object.freeze(mergeObj(a, b));

/**
 * @category Dictionary
 */
export const nextDict = new NextCache<TDictionaryRoot, TDictionaryRoot>(merge);

/**
 * @category Dictionary
 */
export const rootDict: TDictionaryRoot = Object.freeze({ $$dict: {} });
