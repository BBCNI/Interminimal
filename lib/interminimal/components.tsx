import {
  ComponentType,
  createContext,
  createElement,
  ReactNode,
  Children,
  useContext,
  Fragment,
  cloneElement,
  isValidElement,
  forwardRef,
  ReactElement
} from "react";

import { parseTemplate } from "./template";
import { LangContextProps, AsType, TProps, TDictionaryRoot } from "./types";
import { LangContext } from "./context";
import { TString } from "./string";

const TContext = createContext(new LangContext());

export const useTranslation = (): LangContext => useContext(TContext);

export const TranslateLocal: ComponentType<
  LangContextProps & { children: ReactNode }
> = ({ children, ...props }) => {
  const ctx = useTranslation().derive(props);
  return <TContext.Provider value={ctx}>{children}</TContext.Provider>;
};

export const Translate: ComponentType<
  LangContextProps & { children: ReactNode; as?: AsType }
> = ({ children, as = "div", ...props }) => {
  const ctx = useTranslation().derive(props);
  return (
    <TText as={as} lang={ctx.language}>
      <TContext.Provider value={ctx}>{children}</TContext.Provider>
    </TText>
  );
};

interface AsProps {
  as: AsType;
  children?: ReactNode;
  [key: string]: any;
}

// Create a component with the specified tag
const As: ComponentType<AsProps> = forwardRef<ReactElement, AsProps>(
  ({ as, children, ...props }, ref) =>
    createElement(as, { ref, ...props }, children)
);

As.displayName = "As";

interface TTextProps {
  children: ReactNode;
  lang: string;
  as: AsType;
  [key: string]: any;
}

export const TText: ComponentType<TTextProps> = forwardRef<
  ReactElement,
  TTextProps
>(({ children, lang, as, ...props }, ref) => {
  const ctx = useTranslation();

  if (lang !== ctx.ambience)
    return (
      <TranslateLocal ambient={lang}>
        <As as={as} ref={ref} {...props} lang={lang}>
          {children}
        </As>
      </TranslateLocal>
    );
  return (
    <As as={as} ref={ref} {...props}>
      {children}
    </As>
  );
});

TText.displayName = "TText";

interface TFormatProps {
  format: string;
  lang: string;
  children?: ReactNode;
}

export const TFormat: ComponentType<TFormatProps> = forwardRef<
  ReactElement,
  TFormatProps
>(({ format, lang, children }, ref) => {
  const ctx = useTranslation();

  const clone = (elt: ReactNode, props?: any): ReactNode =>
    isValidElement(elt) ? cloneElement(elt, props) : elt;

  const parts = parseTemplate(format);

  // Bail out quickly in the simple case
  if (parts.length === 1 && typeof parts[0] === "string")
    return <Fragment>{parts[0]}</Fragment>;

  // Make children into a regular array of nodes
  const params = Children.map<ReactNode, any>(children, x => x) || [];

  if (process.env.NODE_ENV !== "production")
    if (ref && params.length !== 1)
      throw new Error(`Can only forward refs to single children`);

  // Set of available indexes
  const avail = new Set(params.map((_x: any, i: number) => i + 1));

  const dict: TDictionaryRoot = { $$dict: {} };

  // Output nodes
  const out = parts.map(part => {
    if (typeof part === "string") return part;
    const { index, name, text } = part;

    if (name && text)
      dict.$$dict[name] = TString.literal(text, lang).dictionary;

    if (index < 1 || index > params.length)
      throw new Error(
        `Arg out of range %${index} (1..${params.length} are valid)`
      );

    if (!avail.has(index)) throw new Error(`Already using arg %${index}`);

    // Mark it used (at least once)
    avail.delete(index);

    if (ref && index === 1) return clone(params[index - 1], { ref });
    return params[index - 1];
  });

  if (process.env.NODE_ENV !== "production")
    if (avail.size) throw new Error(`Unused args: ${avail}`);

  if (Object.keys(dict.$$dict).length)
    return <TranslateLocal dictionary={dict}>{out}</TranslateLocal>;

  return <Fragment>{out}</Fragment>;
});

TFormat.displayName = "TFormat";

export const T: ComponentType<TProps> = forwardRef<ReactElement, TProps>(
  ({ children, tag, text, content, count, as = "span", ...props }, ref) => {
    const ctx = useTranslation();

    if (content) {
      if (tag || text)
        throw new Error(`Please don't mix content with tag or text`);
      const ts = ctx.translate(content);
      return (
        <TText
          as={as}
          ref={ref}
          lang={ts.language}
          {...ctx.resolveMagicProps(props, ts.language)}
        >
          {ts.toString(count)}
        </TText>
      );
    }

    if (tag || text) {
      const ts = ctx.resolveTranslationProps(tag, text);

      return (
        <TText
          as={as}
          lang={ts.language}
          {...ctx.resolveMagicProps(props, ts.language)}
        >
          <TFormat lang={ts.language} format={ctx.render(ts, count)}>
            {children}
          </TFormat>
        </TText>
      );
    }

    return (
      <TText
        as={as}
        ref={ref}
        lang={ctx.defaultLang}
        {...ctx.resolveMagicProps(props)}
      >
        {children}
      </TText>
    );
  }
);

T.displayName = "T";

const boundMap = new Map();

export const tBind = (as: AsType): ComponentType<TProps> => {
  const bind = (as: AsType): ComponentType<TProps> => {
    const bound: ComponentType<TProps> = forwardRef(
      ({ children, ...props }, ref) => (
        <T as={as} ref={ref} {...props}>
          {children}
        </T>
      )
    );
    const asName =
      typeof as === "string" ? as : as.displayName ?? as.name ?? "anon";
    bound.displayName = `T${asName}`;
    Object.defineProperty(bound, "name", { value: bound.displayName });
    return bound;
  };

  let bound = boundMap.get(as);
  if (!bound) boundMap.set(as, (bound = bind(as)));
  return bound;
};

export const tBindMulti = (as: AsType[]) => as.map(tBind);
