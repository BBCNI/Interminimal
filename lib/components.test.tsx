import React, { ComponentType, FunctionComponent, ReactNode, Ref } from "react";
import R from "react-test-renderer";

import { Translate, T, tBind, tBindMulti } from "./components";
import { TDictionaryRoot } from "./dictionary";

const dictionary: TDictionaryRoot = {
  $$dict: {
    site: { en: "Interminimal" },
    heading: {
      en: "Let's Translate!",
      fr: "Traduisons!",
      de: "Lassen Sie uns Ã¼bersetzen!"
    },
    maybe: { $$dict: { site: { en: "Something else" } } }
  }
};

interface MyLinkProps {
  ref?: Ref<string>;
  href: string;
  refValue: string;
  children: ReactNode;
}

const MyLink: ComponentType<MyLinkProps> = React.forwardRef<
  string,
  MyLinkProps
>(({ children, href, refValue }, ref) => {
  // @ts-ignore
  if (ref) ref.current = refValue;
  return <a href={href}>{children}</a>;
});

beforeEach(() => {
  jest.spyOn(console, "error");
  // @ts-ignore jest.spyOn adds this functionallity
  console.error.mockImplementation(() => null);
});

afterEach(() => {
  // @ts-ignore jest.spyOn adds this functionallity
  console.error.mockRestore();
});

describe("Interminimal Components", () => {
  it("should render <Translate>", () => {
    expect(
      R.create(
        <Translate lang="fr">
          <T text="Hello" />
        </Translate>
      ).toJSON()
    ).toMatchSnapshot();

    expect(
      R.create(
        <Translate lang="en">
          <T text="Hello" />
        </Translate>
      ).toJSON()
    ).toMatchSnapshot();
  });

  it("should render templates", () => {
    expect(
      R.create(
        <Translate lang="en">
          <T text="Second: %2, first: %1">
            <T text="One" />
            <T text="Two" />
          </T>
        </Translate>
      ).toJSON()
    ).toMatchSnapshot();
  });

  it("should render content", () => {
    expect(
      R.create(
        <Translate lang="en">
          <T content="Second: %2, first: %1" />
        </Translate>
      ).toJSON()
    ).toMatchSnapshot();
  });

  it("should allow element of Translate to be overridden", () => {
    expect(
      R.create(
        <Translate as="p" lang="en">
          <T content="Hello!" />
        </Translate>
      ).toJSON()
    ).toMatchSnapshot();
  });

  it("should bind T to a tag", () => {
    const [Tp, Ti] = tBindMulti(["p", "i"]);
    expect(
      R.create(
        <Translate lang="en">
          <Tp text="para" />
          <Ti text="italic" />
        </Translate>
      ).toJSON()
    ).toMatchSnapshot();
  });

  it("should bind T as a component", () => {
    const TMyLink = tBind(MyLink as FunctionComponent);
    expect(
      R.create(
        <Translate lang="en">
          <TMyLink href="/" text="Link" />
        </Translate>
      ).toJSON()
    ).toMatchSnapshot();
  });

  it("should cache bound components", () => {
    const Tp1 = tBind("p");
    const Tp2 = tBind("p");
    expect(Tp2).toBe(Tp1);
  });

  it("should resolve tags", () => {
    expect(
      R.create(
        <Translate lang="en" dictionary={dictionary}>
          <T tag="heading" />
          <Translate lang="fr">
            <T tag="heading" />
          </Translate>
        </Translate>
      ).toJSON()
    ).toMatchSnapshot();
  });

  it("should translate properties", () => {
    expect(
      R.create(
        <Translate lang="en" dictionary={dictionary}>
          <T t-title={["heading"]}>Some text</T>
        </Translate>
      ).toJSON()
    ).toMatchSnapshot();
  });

  it("should pass template fragments down", () => {
    const ts = { en: "Here is a %1[link] and this %2[is italic]" };
    expect(
      R.create(
        <Translate lang="en">
          <T text={ts}>
            <T as="a" href="/" tag="%1" />
            <T as="i" tag="%2" />
          </T>
        </Translate>
      ).toJSON()
    ).toMatchSnapshot();
  });

  it("should pass refs via Format", () => {
    const ref = React.createRef<boolean>();

    // Test that MyLink gets a ref even though there's no explicit
    // ref in its properties.
    expect(
      R.create(
        <Translate lang="en">
          <T ref={ref} text="[%1]">
            <MyLink refValue="foo" href="/">
              Link
            </MyLink>
          </T>
        </Translate>
      ).toJSON()
    ).toMatchSnapshot();
    expect(ref.current).toBe("foo");
  });

  it("should handle ambience", () => {
    const dict = {
      $$dict: {
        info: { "en-GB": "Spiffing! %1", "en": "Good %1" },
        switch: { en: "switch %1" },
        caption: { "en-GB": "Tally Ho!", "en": "Let's go" }
      }
    };

    expect(
      R.create(
        <Translate lang="en-GB" dictionary={dict}>
          <T as="p" tag="info">
            <T tag="switch">
              <T tag="caption" />
            </T>
          </T>
        </Translate>
      )
    ).toMatchSnapshot();
  });

  it("should retain ambience", () => {
    const dict = {
      $$dict: {
        info: { "en-GB": "Spiffing! %1", "en": "Good %1" },
        switch: { en: "switch %1" },
        caption: { "en-GB": "Tally Ho!", "en": "Let's go" }
      }
    };

    expect(
      R.create(
        <Translate lang="en-GB" dictionary={dict} retainAmbience={true}>
          <T as="p" tag="info">
            <T tag="switch">
              <T tag="caption" />
            </T>
          </T>
        </Translate>
      )
    ).toMatchSnapshot();
  });

  ////////////////////
  // Negative cases //
  ////////////////////

  it("should fail for indexes out of range", () => {
    expect(() =>
      R.create(
        <Translate lang="en">
          <T text="%0">
            <T as="a" href="/" tag="%1" />
          </T>
        </Translate>
      )
    ).toThrow(/out of range/);
  });

  it("should fail for unused children", () => {
    expect(() =>
      R.create(
        <Translate lang="en">
          <T text="%1">
            <span />
            <span />
          </T>
        </Translate>
      )
    ).toThrow(/unused args/i);
  });

  it("should fail for reused children", () => {
    expect(() =>
      R.create(
        <Translate lang="en">
          <T text="%1%1">
            <span />
          </T>
        </Translate>
      )
    ).toThrow(/already/i);
  });

  it("should fail if tag + text provided", () => {
    expect(() =>
      R.create(
        <Translate lang="en">
          <T tag="foo" text="bar" />
        </Translate>
      )
    ).toThrow(/both/);
  });

  it("should fail if tag + content provided", () => {
    expect(() =>
      R.create(
        <Translate lang="en">
          <T tag="foo" content="bar" />
        </Translate>
      )
    ).toThrow(/don't mix/);
  });

  it("should fail to pass refs to multiple children", () => {
    const ref = React.createRef();
    expect(() =>
      R.create(
        <Translate lang="en">
          <T ref={ref} text="%1%2">
            <T as="a" href="/" text="foo" />
            <T as="i" text="bar" />
          </T>
        </Translate>
      )
    ).toThrow(/only forward/);
  });

  it("should fail to pass refs to non-elements", () => {
    const ref = React.createRef();
    expect(() =>
      R.create(
        <Translate lang="en">
          <T ref={ref} text="%1%2">
            Hello!
          </T>
        </Translate>
      )
    ).toThrow(/non-element/i);
  });

  it("should fail to pass refs to content", () => {
    const ref = React.createRef();
    expect(() =>
      R.create(
        <Translate lang="en">
          <T ref={ref} content="Hello" />
        </Translate>
      )
    ).toThrow(/pass ref/i);
  });

  it("should fail to pass refs to regular children", () => {
    const ref = React.createRef();
    expect(() =>
      R.create(
        <Translate lang="en">
          <T ref={ref}>Hello</T>
        </Translate>
      )
    ).toThrow(/pass ref/i);
  });
});
