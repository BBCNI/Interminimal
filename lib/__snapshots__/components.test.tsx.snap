// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[
  `Interminimal Components should allow element of Translate to be overridden 1`
] = `
<p>
  <span>
    Hello!
  </span>
</p>
`;

exports[`Interminimal Components should bind T as a component 1`] = `
<div>
  <a
    href="/"
  >
    Link
  </a>
</div>
`;

exports[`Interminimal Components should bind T to a tag 1`] = `
<div>
  <p>
    para
  </p>
  <i>
    italic
  </i>
</div>
`;

exports[`Interminimal Components should pass refs via Format 1`] = `
<div>
  <span>
    [
    <a
      href="/"
    >
      Link
    </a>
    ]
  </span>
</div>
`;

exports[`Interminimal Components should pass template fragments down 1`] = `
<div>
  <span>
    Here is a 
    <a
      href="/"
    >
      link
    </a>
     and this 
    <i>
      is italic
    </i>
  </span>
</div>
`;

exports[`Interminimal Components should render <As> 1`] = `<div />`;

exports[`Interminimal Components should render <As> 2`] = `<h1 />`;

exports[`Interminimal Components should render <Translate> 1`] = `
<div
  lang="fr"
>
  <span
    lang="en"
  >
    Hello
  </span>
</div>
`;

exports[`Interminimal Components should render <Translate> 2`] = `
<div>
  <span>
    Hello
  </span>
</div>
`;

exports[`Interminimal Components should render content 1`] = `
<div>
  <span>
    Second: %2, first: %1
  </span>
</div>
`;

exports[`Interminimal Components should render templates 1`] = `
<div>
  <span>
    Second: 
    <span>
      Two
    </span>
    , first: 
    <span>
      One
    </span>
  </span>
</div>
`;

exports[`Interminimal Components should resolve tags 1`] = `
<div>
  <span>
    Let's Translate!
  </span>
  <div
    lang="fr"
  >
    <span>
      Traduisons!
    </span>
  </div>
</div>
`;

exports[`Interminimal Components should translate properties 1`] = `
<div>
  <span
    title="Let's Translate!"
  >
    Some text
  </span>
</div>
`;
